Sources/PackageGraph/PackageGraph+Loading.swift

```diff
@@ -802,20 +803,67 @@ private class ModuleAliasTracker {
+               let directTargets = productToDirectTargets[product] {
+                // Keep track of all targets in this product that directly
+                // or indirectly depend on the child product
+                let directRelevantTargets = directTargets.filter {$0.dependsOn(product: child)}
+                var relevantTargets = directTargets.map{$0.dependentTargets()}.flatMap{$0}.filter {$0.dependsOn(product: child)}
+                relevantTargets.append(contentsOf: directTargets)
+                relevantTargets.append(contentsOf: directRelevantTargets)
+                let relevantTargetSet = Set(relevantTargets)
+
+                // Used to compare with aliases defined in other child products
+                // and detect a conflict if any.
+                let allTargetsInOtherChildProducts = childProducts.filter{$0 != child }.compactMap{productToAllTargets[$0]}.flatMap{$0}
+                let allTargetNamesInChildProduct = productToAllTargets[child]?.map{$0.name} ?? []
+                for curTarget in relevantTargetSet {
+                    for (nameToBeAliased, aliasInChild) in childTargetsAliases {
+                        // If there are targets in other child products that
+                        // have the same name that's being aliased here, but
+                        // targets in this child product don't, we need to use
+                        // alias values of those targets as they take a precedence
+                        let otherAliasesInChildProducts = allTargetsInOtherChildProducts.filter{$0.name == nameToBeAliased}.compactMap{$0.moduleAliases}.flatMap{$0}.filter{$0.key == nameToBeAliased}
+                        if !otherAliasesInChildProducts.isEmpty,
+                           !allTargetNamesInChildProduct.contains(curTarget.name) {
+                            for (aliasKey, aliasValue) in otherAliasesInChildProducts {
+                                // Reset the old alias value with this aliasValue
+                                if curTarget.moduleAliases?[aliasKey] != aliasValue {
+                                    curTarget.addModuleAlias(for: aliasKey, as: aliasValue)
+                                }
+                            }
+                        } else {
+                            // If there are multiple aliases for the same key from
+                            // different child products, and their values are
+                            // different, aliasing should not be applied.
+                            let hasMultiValuesForAliasing = allTargetsInOtherChildProducts.compactMap{$0.moduleAliases}.flatMap{$0}.contains{$0.key == nameToBeAliased && $0.value != aliasInChild}
+                            if hasMultiValuesForAliasing {
+                                // If there are aliases, remove as aliasing should
+                                // not be applied
+                                curTarget.removeModuleAlias(for: nameToBeAliased)
```

Comment:
Is this something that should lead to a user-facing diagnostic?

Reply:
yeah, will track it in a separate radar for this.